File: core/src/main/java/org/maibot/core/util/TimerProxy.java
package org.maibot.core.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.function.Supplier;

@SuppressWarnings({"unused", "UnusedReturnValue"})
public class TimerProxy {
    public static void start(Runnable task, String format, Logger log) {
        long startTime = System.currentTimeMillis();
        task.run();
        long endTime = System.currentTimeMillis();
        log.debug(format, (endTime - startTime));
    }

    public static void start(Runnable task, String format, String loggerName) {
        start(task, format, LoggerFactory.getLogger(loggerName));
    }

    public static void start(Runnable task, String format) {
        start(task, format, LoggerFactory.getLogger("Timer"));
    }

    public static void start(Runnable task) {
        start(task, "Task executed in {} ms", LoggerFactory.getLogger("Timer"));
    }

    public static <T> T start(Supplier<T> task, String format, Logger log) {
        long startTime = System.currentTimeMillis();
        var ret = task.get();
        long endTime = System.currentTimeMillis();
        log.debug(format, (endTime - startTime));
        return ret;
    }

    public static <T> T start(Supplier<T> task, String format, String loggerName) {
        return start(task, format, LoggerFactory.getLogger(loggerName));
    }

    public static <T> T start(Supplier<T> task, String format) {
        return start(task, format, LoggerFactory.getLogger("Timer"));
    }

    public static <T> T start(Supplier<T> task) {
        return start(task, "Task executed in {} ms", LoggerFactory.getLogger("Timer"));
    }
}

File: core/src/main/java/org/maibot/core/util/TimerProxy.kt
package org.maibot.core.util

import org.slf4j.Logger
import org.slf4j.LoggerFactory

@Suppress("unused")
object TimerProxy {
    fun <T> start(format: String = "Task executed in {} ms", loggerName: String, task: () -> T): T {
        return start(format, LoggerFactory.getLogger(loggerName), task)
    }

    fun <T> start(
        format: String = "Task executed in {} ms",
        log: Logger = LoggerFactory.getLogger("Timer"),
        task: () -> T
    ): T {
        val startTime = System.currentTimeMillis()
        val ret = task.invoke()
        val endTime = System.currentTimeMillis()
        log.debug(format, (endTime - startTime))
        return ret
    }
}
